### ISSUE TITLE
Action executor runs once every 3 steps

### ISSUE THREAD

**Issue by FIIFE** [caesuric/mountain-goap#4]
Hello,

First of all thanks for sharing your hard work @caesuric. Out of all the alternatives I've looked up this one seemed to have the most intuitive API for me. I think it's great and I hope to see it grow!

So regarding my problem.
I'm trying to implement mountain goap in my real time diablo like game in Unity.
I want my agent to follow the player once he enters agent's vision. I've got goals, sensors, actions and state setup and I'm executing agent.Step() on every update as documentation suggests.

And It works but there is one issue. The executor is fired once every 3 steps which results in choppy inaccurate movement because my movement executor is meant to be run on every update/step.

Have you encountered this issue before? I'd appreciate any help. Thanks!

**Comment by caesuric** [comment]
I have encountered something like this, yes. One way to handle this is to make your movement executor fire a coroutine that executes smooth movement every frame. I find it helpful to abstract actions a bit to accommodate this kind of thinking. So for instance, in the first game I used something like this system in, which is also a Diablo-like Unity game, I had a `MoveToEnemy` action which controls a NavMeshAgent to move the monster towards its target. The same thing could be accomplished with a coroutine if you'd rather not use NavMeshAgents.

I did have some noticeable choppiness with the "focusing" decision process. So the enemy would move smoothly towards the player but would pause and reconsider for a handful of frames when the player moved too much. This is a result of how the Step loop is currently implemented. If you like, I can try to address that for situations where you need an action to execute on every step. The best way to fix this for good would probably be to make a "continue action" callback that keeps executing until the action is done and runs on every step, and to have a better concept of actions that are mid-execution.

Also happy to give examples if that would be helpful. :)

**Comment by FIIFE** [comment]
I'm having trouble figuring out how this coroutine approach could look like so I'd love an example if you don't mind.
I'm also using NavMeshAgent btw.

Executing an action on every step is how i expected it to work so it would definitely be more intuitive for a newbie like me :)
If you think that's reasonable and applicable beyond the scope of my case then i'd appreciate it if you tried to adress it.

Thanks!

**Comment by FIIFE** [comment]
No need to apologize for the wall of text, I appreciate it very much!

So at first I tried solving the issue with your second suggestion using `agent.Step(mode: StepMode.OneAction)` but with no luck. The action still executed once every 3 steps. Maybe my implementation is not right, I'm not sure.

But I succeeded implementing smooth movement on every step using your first suggestion with "movingToPlayer"!

I still encountered some problems along the way while implementing this approach but it's out of scope of this issue so I created new ones:
- https://github.com/caesuric/mountain-goap/issues/7
- https://github.com/caesuric/mountain-goap/issues/8

I don't think there's anything left to do here. I've found a way to implement smooth movement and I saw that you've created the [issue to create ContinueAction feature](https://github.com/caesuric/mountain-goap/issues/5) so I'm closing this.

Thanks for your commitment @caesuric!
